### PR TITLE
gh-107298: Fix C API Buffer documentation

### PR DESCRIPTION
**Pull request by vstinner**
* Issue: gh-107298

----
:books: Documentation preview :books:: https://cpython-previews--108011.org.readthedocs.build/

### CODE REVIEW

**Comment by vstinner** [comment]
Fix some of the Sphinx warnings reported in https://github.com/python/cpython/pull/108010/files PR (PR unrelated to documentation).

File: Doc/c-api/buffer.rst [hunk 1]
```diff
@@ -444,7 +448,7 @@ Buffer-related functions
 
    Send a request to *exporter* to fill in *view* as specified by  *flags*.
    If the exporter cannot provide a buffer of the exact type, it MUST raise
-   :c:data:`PyExc_BufferError`, set ``view->obj`` to ``NULL`` and
+   :exc:`BufferError`, set ``view->obj`` to ``NULL`` and
```
Comment (serhiy-storchaka) [inline Doc/c-api/buffer.rst:451]: There is yet one reference in typeobj.rst.

File: Doc/c-api/buffer.rst [hunk 2]
```diff
@@ -163,13 +163,6 @@ a buffer, see :c:func:`PyObject_GetBuffer`.
       and :c:member:`~Py_buffer.suboffsets` MUST be ``NULL``.
       The maximum number of dimensions is given by :c:macro:`PyBUF_MAX_NDIM`.
 
-   .. :c:macro:: PyBUF_MAX_NDIM
```
Comment (serhiy-storchaka) [inline Doc/c-api/buffer.rst:166]: You can just add
```
   .. c:namespace:: NULL
```
here, and restore the namespace after the description to fix reference issues.

I already used it for some macros declared not at the top level.

https://www.sphinx-doc.org/en/master/usage/restructuredtext/domains.html#namespacing

Comment (vstinner) [reply Doc/c-api/buffer.rst:166]: I saw your "namespace: NULL" trick of your other PRs, but here Py_buffer members are defined before **and** after this macro which is **not** part of Py_buffer structure. In terms of Sphinx, it's confusing. I prefer to document it somewhere else.

**Review by serhiy-storchaka — APPROVED** [review]

**Comment by vstinner** [comment]
Thanks for the review @serhiy-storchaka.

**Comment by serhiy-storchaka** [comment]
Conflict in 3.12 but not in 3.11?

**Pull request merged**